### PR TITLE
WRO-6288:Docs: Apply gatsby-transformer-documentationjs plugin to extract JSDoc tags directly

### PR DESCRIPTION
**Pull request by sjro**
### Checklist

* [x] I have read and understand the [contribution guide](http://enactjs.com/docs/developer-guide/contributing/)
* [ ] A [CHANGELOG entry](http://enactjs.com/docs/developer-guide/contributing/changelogs/) is included
* [ ] At least one test case is included for this feature or bug fix
* [x] Documentation was added or is not needed
* [ ] This is an API breaking change

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)
Want to skip the process of converting jsdocs to json.

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)
1. Use gatsby-transformer-documentationjs plugin (https://www.gatsbyjs.com/plugins/gatsby-transformer-documentationjs/)
2. `gatsby-remark-prismjs` should be placed after `gatsby-remark-autolink-headers` (https://www.gatsbyjs.com/plugins/gatsby-remark-autolink-headers)
3. Pages are required as many as the number of @module tags. So, in the onCreateNode function, among the documentationjs nodes, only nodes with module tag are created as new nodes by setting the type to ApiDoc.
4. 'CreatePages' is called as many as the number of newly created nodes (AllApiDoc) in this way.
5. Connect to localhost:8000__graphql and extract the necessary data keys
6. Add jsdocs-copy command to copy jsdocs files  under src/jsdocs/docs/modules directory.
7. Implementing the jsdocs-copy command using git sparse (https://git-scm.com/docs/git-sparse-checkout)

### Additional Considerations
[//]: # (How should the change be tested?)
[//]: # (Are there any outstanding questions?)
[//]: # (Were any side-effects caused by the change?)

- Temporarily upload the docIndex.json to git before implementing search function. 

- Git version of test machine must be 2.25 or higher

### Links
[//]: # (Related issues, references)
WRO-6288, WRO-7524

### Comments
Enact-DCO-1.0-Signed-off-by: SJ RO (sj.ro@lge.com)

### CODE REVIEW

File: gatsby-node.js [hunk 1]
```diff
@@ -84,55 +84,80 @@ exports.onCreateBabelConfig = ({actions}) => {
...
 	const {createNodeField, createNode, createParentChildLink} = actions;
 	let slug;
 	if (node.internal.type === 'MarkdownRemark') {
 		const fileNode = getNode(node.parent);
-		slug = createSlug(fileNode);
+		if(fileNode.internal.type === 'File') {
+			slug = createSlug(fileNode);
 
+			// Add slug as a field on the node.
+			createNodeField({node, name: 'slug', value: slug});
```
Comment (juwonjeong) [inline gatsby-node.js:116]: As I checked, if fileNode.internal.type is 'File', createNodeField will be called twice (in L116 and L121). Is it designed to called twice?

File: gatsby-node.js [hunk 2]
```diff
@@ -157,8 +182,8 @@ exports.createPages = ({graphql, actions}) => {
 	}
 
 	return new Promise((resolve, reject) => {
-		const markdownPage = path.resolve('src/templates/markdown.js');
-		const jsonPage = path.resolve('src/templates/json.js');
+		const markdownPage = path.resolve('./src/templates/markdown.js');
+		const apiPage = path.resolve('./src/templates/apidoc.js');
```
Comment (juwonjeong) [inline gatsby-node.js:186]: How about apiDocPage to be more specific?
```suggestion
		const apiDocPage = path.resolve('./src/templates/apidoc.js');
```

File: scripts/prepareRaw.js [hunk 3]
```diff
@@ -35,13 +36,68 @@ function copyGitHub (repo, destination, force, branch = 'master', useSSH) {
...
 const args = parseArgs(process.argv);
 const rebuild = args['rebuild-raw'],
-	extraRepos = args['extra-repos'];
+	extraRepos = args['extra-repos'],
+	jsdocsCopy = args.jsdocs;
 
 copyGitHub('enactjs/enact', 'raw/enact', rebuild, args['enact-branch'], args['ssh']);
 copyGitHub('enactjs/cli', 'raw/cli', rebuild, args['cli-branch'], args['ssh']);
 copyGitHub('enactjs/eslint-config-enact', 'raw/eslint-config-enact', rebuild, args['eslint-config-branch'], args['ssh']);
+if(jsdocsCopy) {
```
Comment (juwonjeong) [inline scripts/prepareRaw.js:98]:
```suggestion
if (jsdocsCopy) {
```

File: scripts/prepareRaw.js [hunk 4]
```diff
@@ -50,6 +106,11 @@ if (extraRepos) {
 		const [name, branch] = repo.split('#'),
 			[, lib] = name.split('/'),
 			dest = `raw/${lib}`;
-		copyGitHub(name, dest, rebuild, branch, args['ssh']);
+
+		if(jsdocsCopy) {
+			copyGitHubParse(name, lib, rebuild, branch, args['ssh']);
+		}else {
```
Comment (juwonjeong) [inline scripts/prepareRaw.js:112]:
```suggestion
		} else {
```

File: scripts/prepareRaw.js [hunk 5]
```diff
@@ -50,6 +106,11 @@ if (extraRepos) {
 		const [name, branch] = repo.split('#'),
 			[, lib] = name.split('/'),
 			dest = `raw/${lib}`;
-		copyGitHub(name, dest, rebuild, branch, args['ssh']);
+
+		if(jsdocsCopy) {
```
Comment (juwonjeong) [inline scripts/prepareRaw.js:110]:
```suggestion
		if (jsdocsCopy) {
```

File: scripts/prepareRaw.js [hunk 6]
```diff
@@ -50,6 +106,11 @@ if (extraRepos) {
...
+			copyGitHubParse(name, lib, rebuild, branch, args['ssh']);
+		}else {
+			copyGitHub(name, dest, rebuild, branch, args['ssh']);
+		}
 	});
-}
+}
```
Comment (juwonjeong) [inline scripts/prepareRaw.js:116]: Please add newline at EOF.

File: src/components/DocParse.js [hunk 7]
```diff
@@ -6,14 +6,29 @@ import css from '../css/main.module.less';
...
+		const text = codeText.replace(/\,/g,'');
+		return <Code type={lang} key={index}>{text}</Code>;
+	}
 }
 
 function parseLink (child, index) {
 	let title = child.children[0].value;
 	const linkText = child.children[0].text || linkReference || title;
-	const url = child.url;
+	const url = (child.properties && child.properties.href);
```
Comment (juwonjeong) [inline src/components/DocParse.js:31]: How about using optional chaining operator?
```suggestion
	const url = child.properties?.href;
```

File: src/components/DocParse.js [hunk 8]
```diff
@@ -41,6 +56,34 @@ function parseLink (child, index) {
 }
 
 function parseChild (child, index) {
+	if(child.type === 'element') {
```
Comment (juwonjeong) [inline src/components/DocParse.js:59]: Please add space between if and '(', and ')' and '{' in line 59 ~ 86.

File: src/utils/functions.js [hunk 9]
```diff
@@ -136,7 +136,7 @@ export const renderExportedFunction = (func) => {
 	const params = func.params || [];
 	const paramStr = buildParamList(params);
 	const name = func.name;
-	const hasReturns = !!func.returns.length;
+	const hasReturns = !!(func.returns && func.returns.length);
```
Comment (juwonjeong) [inline src/utils/functions.js:139]:
```suggestion
	const hasReturns = !!(func.returns?.length);
```
And please also fix L161

**Comment by juwonjeong** [comment]
I think it would be better to add src/jsdocs folder in .gitignore file, because hundreds of files are created after jsdocs-copy command.

File: gatsby-node.js (hunk 1 above)
Comment (juwonjeong) [reply gatsby-node.js]: It's fixed.

File: gatsby-node.js (hunk 2 above)
Comment (juwonjeong) [reply gatsby-node.js]: It's fixed.

**Comment by juwonjeong** [comment]
After building enact-docs, I found that some information was missing from the page.
Please check pages below. (I tried to compare with enatjs.com pages)
1) core/dispatcher 
- off, on
In param target, `default: 'document'` is missed.
- once
In param target, `optional` is missed.

2) core/handle
- forward, log, preventDefault, stop
`-> true` is missed.

-EventAdatper, HandlerFunction 
props, context parameter written with wrong type(`of`)

3) core/keymap
- add, remove
keyCode parameter variableTypesKey is missed
- addAll
set parameter written with wrong type(`of`)

4) core/kind
- kind
please check code block
`kind( config ) -> of`
- StylesBlock
publicClassNames has empty variableTypesKey

5) Sandstone API
Properties didn't have variableTypesKey like String, Object, Boolean.

**Comment by juwonjeong** [comment]
In addition, In sandstone/TabLayout module, `statics` field is omitted when compared to the enactjs.com site.

File: src/utils/functions.js (hunk 9 above)
Comment (sjro) [reply src/utils/functions.js]: fixed it

**Comment by juwonjeong** [comment]
When we visit Developer Guide page and Developer Tools page,
There are some unnecessary links in the index section and links are not working properly. Please check developer guide menu and developer tools page.

**Comment by juwonjeong** [comment]
LGTM

**Comment by seunghoh** [comment]
@sjro Hm... there are lots of space mistakes in the `if` statement. PR should be ZERO error on strict linting. You should run `npm run lint` to make sure no linting error before PR.

**Pull request merged**